### PR TITLE
Runner Cleanup: SuiteRunner & TestRunner classes

### PR DESCRIPTION
**Pull request by flashdesignory**
- move SuiteRunner class into a separate file
- create a new TestRunner class, which a remote workload could consume in the future
- rename _runTestAndRecordResults to TestRunner.runTest()
- create class methods for functions used in the runSync & measureAsync functions

### CODE REVIEW

**Comment by bgrins** [comment]
I like the direction here - would like @julienw to review the change if possible.

@flashdesignory to confirm - the big picture idea here is that remote tests will (eventually) consume resources/test-runner.mjs themselves, and it will grow functionality to manage communication back and forth with the suite (postMessage etc)?

**Comment by flashdesignory** [comment]
> @flashdesignory to confirm - the big picture idea here is that remote tests will (eventually) consume resources/test-runner.mjs themselves, and it will grow functionality to manage communication back and forth with the suite (postMessage etc)?

remote tests would use the test-runner class, that's correct 😄.
postMessage communication will happen on a higher level in the workloads, not on a test-runner level.

File: resources/test-runner.mjs [hunk 1]
```diff
@@ -0,0 +1,96 @@
+import { TEST_INVOKER_LOOKUP } from "./test-invoker.mjs";
+
+export class TestRunner {
+    constructor(suite, test, params, callback, page, frame) {
```
Comment (rniwa) [inline resources/test-runner.mjs:4]: Can we use a consistent parameter order as SuiteRunner? so "frame, page, params"

File: resources/test-runner.mjs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        this._suite = suite;
```
Comment (rniwa) [inline resources/test-runner.mjs:5]: We should use real private instance variables.

File: resources/test-runner.mjs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        };
+
+        const report = () => this._callback(this._test, syncTime, asyncTime);
+        const invokerClass = TEST_INVOKER_LOOKUP[this._params.measurementMethod];
+        const invoker = new invokerClass(runSync, measureAsync, report, this._params);
+
+        return invoker.start();
+    }
+
+    _runWarmupSuite() {
```
Comment (rniwa) [inline resources/test-runner.mjs:44]: This isn't really a suite. We should probably just call it _runWarmup or something

File: resources/test-runner.mjs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        const syncEndLabel = `${suiteName}.${testName}-sync-end`;
+        const asyncStartLabel = `${suiteName}.${testName}-async-start`;
+        const asyncEndLabel = `${suiteName}.${testName}-async-end`;
+
+        let syncTime;
+        let asyncStartTime;
+        let asyncTime;
+        const runSync = () => {
+            this._runWarmupSuite();
+            syncTime = this._measureSyncTime(syncStartLabel, syncEndLabel);
```
Comment (rniwa) [inline resources/test-runner.mjs:28]: This code is harder to follow with nested function calls. We should go back to having all the code here instead.

File: resources/test-runner.mjs [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+            asyncStartTime = this._initAsyncTime(asyncStartLabel);
+        };
+        const measureAsync = () => {
+            this._forceLayout();
```
Comment (rniwa) [inline resources/test-runner.mjs:32]: Ditto.

File: resources/test-runner.mjs (hunk 4 above)
Comment (flashdesignory) [reply resources/test-runner.mjs:28]: I was trying to find a good way to be able to reuse as much as possible for remote workloads. 
If I keep everything in one function, I'll end up overriding the whole function, when I might just need to change one line in one nested function call. 

With this, I should be able to create a new class that extends TestRunner, but only overrides one method, which is more targeted and introduces less CLs.

Comment (rniwa) [reply resources/test-runner.mjs:28]: Which is the line that needs to be changed for remote workloads?

Comment (flashdesignory) [reply resources/test-runner.mjs:28]: Some examples, that are slightly different for remote workloads:

`this._test.run(this._page); `
The remote workloads don't use a page object to wrap elements. 
Also, if we opt into async workloads, we might want to await the run function. 

`_forceLayout()`
It works, although I won't use or pass in a frame reference.

These are just some examples

Comment (flashdesignory) [reply resources/test-runner.mjs:28]: Both examples can work as is (even if page and frame are not passed in), but I still feel that smaller methods might be easier to follow and reuse. 

This shouldn't be a blocker though and if I'm the only one feeling this way, I can certainly change it.

Comment (bgrins) [reply resources/test-runner.mjs:28]: Perhaps we should make it more of a copy/paste for this PR and then follow up with refactors? I think it'll be easier to say one way or the other with (a) the change unbundled from the file move and (b) when we have a remote implementation to consider.

Comment (rniwa) [reply resources/test-runner.mjs:28]: Yeah, doing the split as a follow up seems like a good idea here.

Comment (flashdesignory) [reply resources/test-runner.mjs:28]: sounds good - i'll fix it up!

File: resources/test-runner.mjs (hunk 1 above)
Comment (flashdesignory) [reply resources/test-runner.mjs]: changed order for both SuiteRunner and TestRunner

File: resources/test-runner.mjs (hunk 2 above)
Comment (flashdesignory) [reply resources/test-runner.mjs]: done

File: resources/test-runner.mjs (hunk 3 above)
Comment (flashdesignory) [reply resources/test-runner.mjs]: done

File: resources/test-runner.mjs (hunk 4 above)
Comment (flashdesignory) [reply resources/test-runner.mjs]: done

File: resources/test-runner.mjs (hunk 5 above)
Comment (flashdesignory) [reply resources/test-runner.mjs]: reverted

**Review by rniwa — APPROVED** [review]

**Review by bgrins — APPROVED** [review]
I'll go ahead and approve in place of Julien with the simplified change in place

**Comment by julienw** [comment]
FTR looks good to me!

**Pull request merged**